documentation/sphinx/source/developer-guide.rst

```diff
@@ -823,3 +823,133 @@ Loading data is a common task in any database. Loading data in FoundationDB will
+* ``too_old`` is thrown if your transaction runs for more than five seconds. If you see this error often, you should try to make your transactions shorter.
+* ``future_version`` is one of the slightly more complex errors. There are several ways how this error could be generated: if you set the read version of your transaction manually to something larger that exists or if the storage servers are falling behind. The second case should be more common. This is usually caused by a write load that is too high for FoundationDB to handle or by faulty/slow disks.
+
+The good thing about these errors is, that retrying is simple: you know that the transaction didn't commit and therefore you can retry even without thinking much about weird corner cases.
+
+The ``unknown_result`` Error
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``unknown_result`` can be thrown during a commit. This error is difficult to handle as you won't know whether your transaction was committed or not. There are mostly two reasons why you might see this error:
+
+#. The client lost the connection to the proxy to which it did send the commit. So it never got a reply and therefore can't know whether the commit was successful or not.
+#. There was a FoundationDB failure - for example a proxy failed during the commit. In that case there is no way for the client know whether the transaction succeeded or not.
+
+However, there is one guarantee FoundationDB gives to the caller: at the point of time where you receive this error, the transaction has either committed or not. Or: it is guaranteed that the transaction is not in-flight anymore. This is an important guarantee as it means that if your transaction is idempotent you can simply retry (see also above ). For more explanations see developer-guide-unknown-results_.
+
+Non-Retryable Errors
+~~~~~~~~~~~~~~~~~~~~
+
+The trickiest errors are non-retryable errors. ``Transaction.on_error`` will rethrow these. Some examples of non-retryable errors are:
+
+#. ``timeout``. If you set a timeout for a transaction, the transaction will throw this error as soon as that timeout accurs.
+#. ``actor_cancelled``. This error is thrown if you call ``cancel()`` on any future returned by a transaction. So if this future is shared by multiple threads or coroutines, all other waiters will see this error.
+
+If you see one of those errors, the best way of action is to fail the client.
+
+At a first glance this looks very similar to an ``unknown_result``. However, for these errors lack the one guarantee ``unknown_result`` still gives to the user: If the commit has already been sent to the database, the transaction could get committed at a later point in time. This means that if you retry the transaction, your new transaction might race with the old transaction. This will cause unexpected results and it does violate FoundationDB's external consistency guarantees.
```

Comment:
as we discussed offline, I think we actually agreed it will violate  external consistency guarantees. One just needs to be carefull about what that guarantee is.

Reply:
I guess maybe the wording is unclear to me. We won't guarantee external consistency in this case, but it may not be violated depending on the circumstances. So maybe rewording it to say that external consistency won't be guaranteed?

Reply:
I think external consistency is still guaranteed. for external consistency one needs to define start and end of the operation/transaction. In the case client times out waiting on a transaction it doesnt know when it ended. The tx can commit any time BEFORE it ended, just the client has no way to know exactly when it ended.
But as long as the commit will not happen after the tx ended, it is still externally consistent.

Reply:
I will try to rewrite this a bit. I think the important bit is that because we abandon the transaction we don't make any guarantees about causality anymore